### PR TITLE
Add in-loop stall detector and live diagnostic tail to History Publish CI fixture

### PR DESCRIPTION
**Pull request by tomerweller**
Refs #3741

## Summary

The `History Publish (Testnet)` CI fixture has failed 4 times (#3280, #3707,
#3727, #3741) with the same "waiting for first published checkpoint" symptom.
Raw CI log investigation (`gh run view --log`, both attempts of #3741's own
failing run) showed both attempts were killed by an **external GitHub Actions
job-level cancellation** — never reaching the script's own internal 1500s
deadline or printing a `DISPOSITION:` line. That external cancellation skips
even `if: always()`-gated steps, so `Upload logs`/`Summary` never ran and
**no validator.log evidence has survived any of the 4 recurrences**.

#3732's prior fix (disabling the fixture's internal watchdog auto-abort)
correctly stopped a still-progressing node from self-SIGABRTing under
legitimate slow-verify load, but it also removed the only code path that
turned "event loop badly stalled" into a clean, script-detected process
death — trading a fast/diagnosable failure for a slow/opaque one.

This PR adds, to `scripts/test-history-publish.sh`:

1. **`is_log_stalled(log_file, now_epoch, threshold_secs)`** — a pure
   mtime-based staleness check on `validator.log`, wired into the poll loop
   (after the existing `kill -0` liveness check) as a hard-red
   `DISPOSITION: STALL` exit (default threshold 180s, overridable via
   `--stall-threshold`). This is **always** a hard red — never covered by
   `--soft-on-sync-timeout` (a stalled-but-alive process is a different
   failure mode than "testnet never became reachable"). Exiting here (rather
   than waiting for the external cancellation) makes the workflow's own
   `if: failure()` Upload-logs step fire, so the full `validator.log`
   artifact is captured.
2. An **incremental, grep-filtered live tail** of new `validator.log` lines
   into the step's own stdout every ~5s poll iteration (filter:
   `WATCHDOG|maxtps_scp|database is locked|straggler timeout|db_write_ctx`,
   the diagnostic families already root-caused for this fixture in #3727 and
   the #3702 `db_write_ctx` instrumentation). GitHub Actions captures step
   stdout even under an external job-level cancellation, unlike the artifact
   upload.
3. A `--classify-stall-only <log_file> <threshold_secs>` offline CLI seam
   mirroring the existing `--classify-only` pattern, so `is_log_stalled()` is
   testable without a live validator/build.

The 180s threshold was chosen against two observed external-cancellation
timings for this fixture (~287s and ~1437s) and against the periodic
progress-logging cadence already present during healthy-but-slow catchup
(bucket-download progress, ledger-prefetch progress, the watchdog's own
15s/30s warn/error tiers) — none of which leave a genuine 180s silent gap.

**This is NOT a fix for the underlying stall.** The root cause is the
SCP-verify/WAL-contention capacity family tracked by #3702/#3266 (confirmed
distinct from #3702's specific "restore from disk" mechanism, since this
fixture always does a fresh cold catchup — see the converged plan for the
full investigation). This PR is diagnostic instrumentation only, so the next
recurrence produces real evidence instead of another artifact-less external
cancellation. #3741 stays open pending that evidence.

## Plan reference

[Converged Plan comment](https://github.com/stellar-experimental/henyey/issues/3741#issuecomment-5057784903)

## Test plan

- [x] `cargo fmt --check`
- [x] `cargo clippy --all -- -D warnings`
- [x] `bash scripts/test-history-publish-harness.sh` — 21/21 TAP assertions pass (14 pre-existing + 7 new)
- [x] `cargo test -p henyey-app --lib config::tests` — 295 passed (unaffected; no config changes)
- [x] `shellcheck scripts/test-history-publish.sh` — same warning count as origin/main (14), no new warnings introduced
- [x] Manual smoke test of `--classify-stall-only` against synthetic stale/fresh fixtures (matches harness assertions)

## Regression test

- **Tests:** `scripts/test-history-publish-harness.sh::test_stall_detector_flags_stale_log`,
  `::test_stall_detector_ignores_fresh_log`, `::test_diagnostic_grep_pins_known_signal_names`
- **Pre-fix:** committed as `e45b49f7` — verified 3 of 7 new assertions FAILED
  (`--classify-stall-only` unrecognized → script hit the `Unknown arg` branch,
  no `DISPOSITION: STALL` marker ever printed, live-tail filter tokens absent
  from the script) — 18/21 total.
- **Post-fix:** verified all 21/21 PASS after `49bd1cc7`.

## Deviations from plan

None — implementation follows the converged plan's file list, approach, and
explicit disposition rule (STALL always hard red) as-is.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/test-history-publish.sh [hunk 1]
```diff
@@ -305,6 +416,34 @@ while true; do
...
+    exit 1
+  fi
+
+  # Incremental live-tail (#3741): echo any new validator.log lines matching
+  # the diagnostic filter into this step's own stdout, which GitHub Actions
+  # captures even under an external job-level cancellation (unlike the
+  # artifact upload). No background process/PID cleanup needed — this stays
+  # inside the existing single-threaded 5s poll loop.
+  CURRENT_LOG_LINES=$(wc -l < "$LOG_FILE" 2>/dev/null || echo 0)
+  if [[ "$CURRENT_LOG_LINES" -gt "$LAST_LOG_LINE" ]]; then
```
Comment (tomerweller) [inline scripts/test-history-publish.sh:442]: **MINOR (edge-case, non-blocking):** small TOCTOU race here — `CURRENT_LOG_LINES` is captured via `wc -l` *before* the `tail -n "+N+1"` read. If `validator.log` grows between those two commands, `tail` will emit lines beyond what `CURRENT_LOG_LINES` recorded, but `LAST_LOG_LINE` is then set to the (now-stale) earlier count — so the next iteration re-tails from the same offset and reprints a few already-seen diagnostic lines. Purely cosmetic (duplicate lines in the live-tail stream, never a functional/disposition bug), but worth a follow-up to snapshot the line count once and reuse it for both the comparison and the tail offset.

File: scripts/test-history-publish.sh [hunk 2]
```diff
@@ -11,13 +11,49 @@
...
+#      (unlike the artifact upload, which does not survive it).
+#   2. Stall-based clean exit: is_log_stalled() checks validator.log's mtime
+#      against --stall-threshold (default 180s). If the node is still alive
+#      (kill -0 succeeds) but the log has produced zero new bytes for that
+#      long, this is treated as a genuine stall — DISPOSITION: STALL, always a
+#      hard red (exit 1, never covered by --soft-on-sync-timeout), so the
+#      workflow's own `if: failure()` Upload-logs step fires and the full
+#      validator.log artifact is captured too.
+#
+# 180s was chosen against two observed external-cancellation timings for this
```
Comment (tomerweller) [inline scripts/test-history-publish.sh:50]: **MINOR (regression-risk, non-blocking):** the converged plan's Critic A explicitly recommended doing one local dry-run of the updated script against live testnet before merging, to sanity-check that no legitimate catchup phase (e.g. a single very large/slow bucket download with no per-chunk progress log) produces a genuine >180s silent gap in validator.log that would false-positive a STALL. The PR's Test Plan checklist shows the offline `--classify-stall-only` synthetic-fixture smoke test but doesn't show this specific live dry-run was performed. Not blocking — this is a strict improvement over the current 0%-diagnostic status quo and the threshold is operator-tunable via `--stall-threshold` without a code change — but recommend watching the next scheduled History Publish run closely for a false-positive STALL disposition.

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** `is_log_stalled()` mtime-staleness logic is sound, the offline `--classify-stall-only` seam calls the exact same function used in the live poll loop (not a parallel reimplementation), and the regression tests genuinely fail before the fix and pass after — verified independently by checking out both commits and running the harness directly.

<details>
<summary>Full review</summary>

**Cycle:** 1 (no prior Correctness comment on this PR).

**Test-verification (kind: bug-fix, per the issue's Triage Report):**
- Checked out the test-only commit `e45b49f7` (harness changes only, no script changes) in an isolated worktree and ran `scripts/test-history-publish-harness.sh` directly: **18/21 pass, 3 fail** — `stale_log_stall_detector_emits_disposition_stall`, `fresh_log_stall_detector_exits_zero`, and `diagnostic_filter_pins_all_known_families` fail exactly as the PR body claims (`--classify-stall-only` unrecognized, script hits the `Unknown arg` branch, no `DISPOSITION: STALL` marker, filter tokens absent). This exactly matches the PR's "## Regression test" section's stated pre-fix count.
- Checked out the fix commit `49bd1cc7` and reran: **21/21 pass, 0 fail.** Matches the PR's claimed post-fix result.
- This confirms genuine failing-test-first discipline — not just a claim in the PR body.

**Correctness of `is_log_stalled(log_file, now_epoch, threshold_secs)`:**
- Missing log file → returns 1 (not stalled); correctly deferred to the pre-existing `kill -0` liveness and HAS-file checks, since `LOG_FILE` is created by shell redirection before the process forks (`> "$LOG_FILE" 2>&1 &`), so in practice the file always exists by the time the poll loop runs — the missing-file branch is a defensive no-op, not a live code path.
- `age = now_epoch - mtime`; `age -ge threshold_secs` → stalled. No off-by-one bug; boundary (`age == threshold`) intentionally counts as stalled, which is the conservative direction.
- No clock-skew concern: both `now_epoch` (`$(date +%s)`) and `mtime` (`stat -c %Y`) come from the same host clock in both the CI runner and the offline harness.
- Confirmed via `grep -n is_log_stalled` that the CLI seam (`--classify-stall-only`) and the live poll-loop call site invoke the identical function — this is a genuine test seam, not a shadow reimplementation that could drift from the real code path.

**Grep filter soundness:** `DIAG_GREP_PATTERN='WATCHDOG|maxtps_scp|database is locked|straggler timeout|db_write_ctx'` is a fixed ERE constant (no untrusted input, no injection risk) run via `grep -E`. `test_diagnostic_grep_pins_known_signal_names` cross-checks these tokens against the actual strings still emitted by `crates/app/src/app/mod.rs` (`WATCHDOG:`) and `crates/app/src/app/lifecycle.rs` (`db_write_ctx = "peer-record-update"`) — verified these strings are still present in current `crates/app` source, so the pin is real, not vacuous.

**Scope verification (explicit ask from the task):** `git diff origin/main pr-head -- .github/workflows/ crates/` is empty — confirmed this PR touches only `scripts/test-history-publish.sh` and `scripts/test-history-publish-harness.sh`. The CI timeout value (1500s, set in `.github/workflows/history-publish.yml`'s `workflow_dispatch` default and passed via `--timeout`) and the watchdog auto-abort logic (`crates/app`, `watchdog_abort_secs`) are both untouched, as required.

**Issue disposition:** confirmed #3741 is still `OPEN` with a "Do: Implementation complete" comment explicitly stating this is instrumentation, not a fix, and that the issue is being left open pending the next recurrence's evidence — matches the task's requirement and avoids repeating the #3280/#3707/#3727 "closed as fixed but wasn't" pattern.

**shellcheck:** ran shellcheck against both origin/main and the PR's `test-history-publish.sh` — identical 14 warnings (all pre-existing `SC2317` false positives on the `trap cleanup EXIT` function), confirming no new warnings, matching the PR's claim.

**Non-blocking (left as inline comments, concern class `edge-case`):** minor TOCTOU race between `wc -l` and the incremental `tail` read in the live-tail loop can cause a few duplicate diagnostic lines across iterations — cosmetic only, never affects the STALL/PUBLISHED/PUBLISH-REGRESSION disposition logic. See inline comment on `scripts/test-history-publish.sh` near the `CURRENT_LOG_LINES` computation.

No correctness blockers found.
</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Risk

**Verdict:** APPROVE

**Summary:** Zero blast radius on the actual validator binary — confirmed the diff touches only two bash CI scripts, no `crates/` or `.github/workflows/` changes. The new STALL disposition changes the *failure mode* of a nightly-only fixture (early clean exit vs. artifact-less external cancellation), which is a strict improvement, not a new regression surface.

<details>
<summary>Full review</summary>

**Cycle:** 1 (no prior Risk comment on this PR).

**Regression risk:** none for production/runtime behavior — `git diff origin/main <pr-head> --stat` shows only `scripts/test-history-publish-harness.sh` and `scripts/test-history-publish.sh` changed (298 lines, +296/-2). No `crates/` Rust source touched, so no XDR/consensus/ledger-observable change, and no risk to the mainnet/testnet validator binary itself. `.github/workflows/history-publish.yml` has no `pull_request:` trigger — this nightly-only cron fixture is never exercised by this PR's own required CI checks, so the only pre-merge validation is the offline TAP harness (verified independently, see Correctness review) plus the next scheduled nightly run.

**Operational/config risk:** `STALL_THRESHOLD_SECS` defaults to 180s and is overridable via `--stall-threshold`; `.github/workflows/history-publish.yml` does not pass `--stall-threshold`, so the nightly job picks up the new default automatically on next merge to `main`. This is an intentional, in-scope behavior change (that's the point of the PR) but worth flagging as a config-risk class: the 180s bound is a considered estimate, not a proven-safe bound, and a false-positive STALL would newly fail a previously-non-failing (if slow) nightly run. The PR's own description and the converged plan both explicitly acknowledge this residual risk and make it operator-tunable without a code change — acceptable given this fixture currently produces **zero** diagnostic value on a real stall (the status quo the PR replaces).

**Security:** no new attack surface — `DIAG_GREP_PATTERN` is a fixed literal, not derived from CLI args or file content; `is_log_stalled` only reads `stat`/mtime, no shell injection vector introduced (no `eval`, no unquoted expansion of file contents into a command).

**API/interface break:** none — both new flags (`--stall-threshold`, `--classify-stall-only`) are additive with safe defaults; existing invocations (including the workflow's own, which doesn't pass either flag) are behaviorally unchanged except for the new STALL disposition path, which was explicitly the point of the change and cannot fire on a healthy run within the 180s bound per the plan's cadence analysis.

**Non-blocking (left as inline comment, concern class `regression-risk`):** the converged plan's Critic A recommended a live testnet dry-run before merge to sanity-check the 180s threshold against real catchup timing; the PR's Test Plan checklist doesn't show this specific live dry-run (only the offline synthetic-fixture smoke test). Not blocking since (a) the threshold is tunable post-merge without a code change, and (b) a false-positive STALL is strictly more informative than today's silent external cancellation — but flagging so the next scheduled run is watched for it.

No risk blockers found.
</details>

**Comment by tomerweller** [comment]
## ✅ Merged

**Commit:** d3cb1f98689420fe18608a7beed29f6280fcf140

Both agent reviewers (Correctness, Risk) verdicted APPROVE on cycle 1. Independently verified: checked out the test-only commit (`e45b49f7`) and confirmed the harness genuinely fails (18/21) before the fix, and the fix commit (`49bd1cc7`) brings it to 21/21; confirmed `shellcheck` warning count is unchanged (14, pre-existing); confirmed the diff touches only `scripts/test-history-publish.sh` and `scripts/test-history-publish-harness.sh` (no `.github/workflows/` or `crates/` changes) — CI timeout value and watchdog logic are untouched, as required. All required CI checks (Test, Doctest, Clippy, Format, Build(release-ci), Analyze(rust)) were green; the only pending check was the non-required testnet Quickstart shard.

**Follow-up issues filed for unaddressed inline review comments:** #3745 (TOCTOU race in live-tail line-count offset — cosmetic, duplicate diagnostic lines), #3746 (confirm 180s stall threshold via a live testnet dry-run)

**Note:** the linked project automation auto-closed #3741 when its board status moved to Done. Per this PR's own explicit design (the converged plan, Critic C, and the PR's "Do" completion comment all state this is diagnostic instrumentation, not a fix for the underlying stall), #3741 has been reopened — it should stay open pending the next recurrence's diagnostic evidence, consistent with #3702/#3266 remaining the owner of the actual capacity fix.
